llama-index-core/llama_index/core/workflow/workflow.py

```diff
@@ -265,6 +267,11 @@ async def _task(
                     # Store the accepted event for the drawing operations
                     ctx._accepted_events.append((name, type(ev).__name__))
 
+                    # Checkpoint
+                    ctx._create_checkpoint(
```

Comment:
so we are automatically creating a checkpoint on every step? This might be unwanted by users, especially since this could potentially eat up a lot of memory and resources.

I would prefer some kind of syntax where checkpointing is more explicit? I was thinking something like `checkpoint_before=["step1", ...]` ? There is also the possibility of passing in a callable that takes in the step name, context, and step event, and returns true/false if we should checkpoint?

Also we don't have to do it this PR, but having some way to store these checkpoints in a db automatically instead of in-memory would probably be useful?

Reply:
Correct me if I'm wrong, but the current UX is "checkpoint everything, filter after" whereas the above is "filter first, checkpoint only what you need"

Reply:
Correct, I personally like this approach better because of the following:
- We can (should) easily disable checkpointing altogether for production deployments
- It seems safer to collect everything and filter later: specially while debugging a long workflow, you don't want to set checkpointing wrong and realize later
- By collecting everything, we don't need to come up with a solution to define checkpoints right now, and that would be difficult to revert/change. This way we can collect some user feedback and pain points for a better UX.

Reply:
Yeah we went with this initial version of checkpointing everything (i.e. every step) cause didn't have a clear picture on how we should enable defining checkpoints that our users would need/want.

I do agree we should make at least turning on/off checkpointing possible at this point.

Reply:
Added `store_checkpoints: bool = True` param in `Workflow.run()` and `Workflow.run_from()` methods. So, turning on and turning off checkpoints is done per run, which I think makes the most sense.

Reply:
Should we default this `store_checkpoints` to `False` instead tho?

Reply:
Fair enough @masci -- I think if store_checkpoints=False by default, I would be happy (I don't want to blow up memory by default heh)